blockchain/chain_sync/src/sync.rs

```diff
@@ -617,48 +616,98 @@ where
 
     /// Validates block semantically according to https://github.com/filecoin-project/specs/blob/6ab401c0b92efb6420c6e198ec387cf56dc86057/validation.md
     async fn validate(&self, block: &Block) -> Result<(), Error> {
+        let mut error_vec: Vec<String> = Vec::new();
+        let mut validations = FuturesUnordered::new();
+
         let header = block.header();
 
         // check if block has been signed
         if header.signature().is_none() {
-            return Err(Error::Validation("Signature is nil in header".to_owned()));
+            error_vec.push("Signature is nil in header".to_owned());
         }
 
         let parent_tipset = self.chain_store.tipset_from_keys(header.parents())?;
 
         // time stamp checks
-        header.validate_timestamps(&parent_tipset)?;
+        if let Err(err) = header.validate_timestamps(&parent_tipset) {
+            error_vec.push(err.to_string());
+        }
+
+        let b = block.clone();
 
+        // Check Block Message and Signatures in them
+        let mut pub_keys = Vec::new();
+        let mut cids = Vec::new();
+        for m in block.bls_msgs() {
+            let pk = self
+                .state_manager
+                .get_bls_public_key(m.from(), parent_tipset.parent_state())?;
+            pub_keys.push(pk);
+            cids.push(m.cid()?.to_bytes());
+        }
```

Comment:
what was the reason for this being done outside of the `check_block_msgs` call? Only reason I mention is that this isn't done async, and would be beneficial if it was also done in the seperate thread

Reply:
this part of code was put outside the `check_block_msgs` call because in order to get it inside, `self.state_manager` would have to be borrowed, and there is no `clone` method that can be used for this, so the setup of `pub_keys` and `cids` wasn't made async.

Reply:
but the function attached to the state_manager only needs the blockstore, so in similar fashion to what you did here you could also do for that? makes sense for this logic to be done inside of check_block_msgs anyway outside of the performance changes